# Exercise 04 - Due Friday, October 21 at 12pm

*Objectives*: Gain experience loading a CSV dataset into a database model you define yourself and using SQL to explore its contents. Explore the data by writing and executing a number of SQL queries using common syntax and functions and describing your findings.

*Grading criteria*: The tasks should all be completed, and questions should all be answered with SQL queries in the space provided, unless a text answer is requested. Results should be correct, according to the embedded tests. The notebook itself should be completely reproducible; from start to finish, another person should be able to use the same code to obtain the same results as yours.  Note that you will receive no more than partial credit if you do not add text/markdown cells explaining your thinking at each major step in each problem.

*Deadline*: Friday, October 21, 12pm. Submit your notebook to Blackboard and push it to your GitHub repository.

## Setup - obtain data and create database

The US City Open Data Census has data on [restaurant inspections from all over the country](http://us-city.census.okfn.org/dataset/food-safety).  Let's take a look at data from Albuquerque.  *Note*: metadata and other details for this dataset are available at http://data.cabq.gov/business/foodinspections/.

Fyi, to save you a step, I've converted this file into a clean CSV file by using the `dos2unix` command to change its line endings and to switch it to use commas instead of tabs using `csvformat -t inputfile.csv > outputfile.csv`.

First we download the dataset:

In [1]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv

--2016-10-24 14:33:19--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.32.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5025226 (4.8M) [text/plain]
Saving to: ‘abq-food-inspections-fy2010.csv’

abq-food-inspection 100%[=====================>]   4.79M  --.-KB/s   in 0.08s  

2016-10-24 14:33:19 (59.8 MB/s) - ‘abq-food-inspections-fy2010.csv’ saved [5025226/5025226]



## Problem 1 - examine the data (35 points)

Use `csvstat` and other tools as you see fit and observe its contents.  Review the metadata definitions linked to above.  Describe what you see:  What do all the columns mean?  Are there null values?  Which columns interest you the most?  Which columns present some complications or questions you would like to ask?

Insert a combination of code and text/markdown cells here to review the data and describe what you find.

First, I use csvstat to take a quick glance at the data.

In [2]:
!shuf -n 10000 abq-food-inspections-fy2010.csv | csvstat

  1. H B M CHILD CARE CENTER
	<class 'str'>
	Nulls: False
	Unique values: 2476
	5 most frequent values:
		SMITHS FOOD AND DRUG CENTERS INC:	89
		ISOTOPES PARK - OVATIONS:	56
		PER EHSII CECELIA GARCIA OOB LA REYNA DEL SUR:	50
		DIONS PIZZA:	40
		CIRCLE K 8940:	39
	Max length: 77
  2. 62708
	<class 'int'>
	Nulls: False
	Min: 13
	Max: 94931
	Sum: 463172920
	Mean: 46321.92419241924
	Median: 49913
	Standard Deviation: 32738.599832710708
	Unique values: 2674
	5 most frequent values:
		46323:	56
		65863:	50
		91821:	39
		93519:	37
		90924:	36
  3. 4311 12TH ST NW 
	<class 'str'>
	Nulls: True
	Unique values: 2344
	5 most frequent values:
		1601 AVENIDA CESAR CHAVEZ SE:	57
		644 OLD COORS BLVD SW:	50
		8601 CENTRAL AV NE:	47
		2200 SUNPORT BLVD SE:	47
		6600 MENAUL BLVD NE:	44
	Max length: 35
  4. ALBUQUERQUE
	<class 'str'>
	Nulls: False
	Values: ALBUQUERQUE
  5. NM
	<class 'str'>
	Nulls: False
	Values: 87, NM
  6. 87107
	<class 'str'>
	Nulls: False
	Unique values: 47
	5 most frequent values:


Then, I exploerd the headings and several rows of the data.

In [3]:
!head abq-food-inspections-fy2010.csv|csvlook

|--------------------------+--------------+-------------------------------+-------------+-------+-------+-----------+-------------------------+--------------------+---------------+-------------+-------------+------------------+------------+------------------+------------------------------+---------------------+-----------------+-------------------------+------------+-------------+-------------+-------------+----------------+----------------+----------------------------------+------------------|
|  FACILITY_NAME           | FACILITY_KEY | SITE_ADDRESS                  | CITY        | STATE | ZIP   | OWNER_KEY | OWNER_NAME              | NATURE_OF_BUSINESS | STREET_NUMBER | STREET_NAME | STREET_TYPE | POST_DIRECTIONAL | PHONE      | PROGRAM_CATEGORY | PROGRAM_CATEGORY_DESCRIPTION | INSPECTION_DATE     | INSPECTION_TYPE | INSPECTION_DESC         | SERIAL_NUM | ACTION_CODE | ACTION_DESC | RESULT_CODE | RESULT_DESC    | VIOLATION_CODE | VIOLATION_DESC                   | INSPECTION_MEMO  |


In [4]:
!csvcut -n abq-food-inspections-fy2010.csv

  1: FACILITY_NAME
  2: FACILITY_KEY
  3: SITE_ADDRESS
  4: CITY
  5: STATE
  6: ZIP
  7: OWNER_KEY
  8: OWNER_NAME
  9: NATURE_OF_BUSINESS
 10: STREET_NUMBER
 11: STREET_NAME
 12: STREET_TYPE
 13: POST_DIRECTIONAL
 14: PHONE
 15: PROGRAM_CATEGORY
 16: PROGRAM_CATEGORY_DESCRIPTION
 17: INSPECTION_DATE
 18: INSPECTION_TYPE
 19: INSPECTION_DESC
 20: SERIAL_NUM
 21: ACTION_CODE
 22: ACTION_DESC
 23: RESULT_CODE
 24: RESULT_DESC
 25: VIOLATION_CODE
 26: VIOLATION_DESC
 27: INSPECTION_MEMO


1.What do all the columns mean? 

There are totally 27 columns.

FACILITY_NAME: Name of the facility where the inspection occurred

FACILITY_KEY: The unique identifier (key) of the facility where the inspection occurred

SITE_ADDRESS: Street address of the facility where the inspection occurred

CITY: City of the facility where the inspection occurred

STATE: State of the facility where the inspection occurred

ZIP:Zip of the facility where the inspection occurred
OWNER_KEY:The unique identifier (key) of the Owner of the facility where the inspection occurred at the time of the inspection

OWNER_NAME: Name of the Owner of the facility where the inspection occurred at the time of the inspection

NATURE_OF_BUSINESS: The type of business, originally for internal use only, provides information about the business. This is not the information provided on the federal tax form

STREET_NUMBER: Street number of the facility where the inspection occurred

STREET_NAME:Street name of the facility where the inspection occurred

STREET_TYPE: Street type of the facility where the inspection occurred

POST_DIRECTIONAL: Postal direction of the facility where the inspection occurred

PHONE:Phone number of the facility where the inspection occurred

PROGRAM_CATEGORY: Id that indicates the type of facility permit the inspection was performed

PROGRAM_CATEGORY_DESCRIPTION: Descript of the id that indicates the type of facility
INSPECTION_DATE: Date the inspection occurred

INSPECTION_TYPE: Code for type of inspection

INSPECTION_DESC: Description of the inspection

SERIAL_NUM: Unique ID for the inspection daily activity. This is a sequential number assigned chronologically as the data is entered into the system

ACTION_CODE: Code of the action taken. If the ACTION_CODE is equal to 00 – NON-GRADED this indicates that the inspector was unable to actually perform an inspection. The RESULT_CODE documents the situation.

ACTION_DESC: Description of the action taken

RESULT_CODE: Code of the result of the inspection

RESULT_DESC: Description of the result of the inspection

VIOLATION_CODE: The code for the violation

VIOLATION_DESC: Description of the violation that was found

INSPECTION_MEMO: Notes from the inspection made by the inspector

2.Are there null values?

There are null values.
For example, there are null values in the columns of INSPECTION_TYPE,INSPECTION_MEMO and other columns.

3.Which columns interest you the most? I am most interested in the column of INSPECTION_TYPE because I wonder why there are many null values in that column.

4.Which columns present some complications or questions you would like to ask? 

Both INSPECTION_TYPE and INSPECTION_MEMO present some complications I would like to ask because there are many null values.


## Problem 2 - define a database model and load the data (35 points)

Based on what you found above, create and connect to a new database, define a database table in it, and load this dataset into it.  You may use either of the methods for this step you have seen in earlier class notebooks.  You may choose to eliminate variables/columns if they are not relevant or interesting to you - explain your reasoning if you do.  Either way, you should load a majority of the columns present in the source dataset into the database and all of its rows.

Once your data has loaded successfully, run a `COUNT(*)` query to verify that all the data has loaded correctly.

Insert a combination of code and text/markdown cells here to connect to your database, define your table, load its data, and verify the loaded records.

I want to know how many lines the csv file has.

In [5]:
!wc -l abq-food-inspections-fy2010.csv

17557 abq-food-inspections-fy2010.csv


I found there are 17557 rows in the csv file. I also rename the csv file.

In [6]:
!mv abq-food-inspections-fy2010.csv food.csv

I loaded sql extensions.

In [7]:
%load_ext sql

/opt/conda/lib/python3.5/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/opt/conda/lib/python3.5/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


Then,I create a database and convert the csv file into the database.

In [8]:
!csvsql --db sqlite:///food.db --insert food.csv

(sqlite3.OperationalError) table food already exists [SQL: '\nCREATE TABLE food (\n\t"FACILITY_NAME" VARCHAR(77) NOT NULL, \n\t"FACILITY_KEY" INTEGER NOT NULL, \n\t"SITE_ADDRESS" VARCHAR(35), \n\t"CITY" VARCHAR(11) NOT NULL, \n\t"STATE" VARCHAR(2) NOT NULL, \n\t"ZIP" VARCHAR(10) NOT NULL, \n\t"OWNER_KEY" INTEGER NOT NULL, \n\t"OWNER_NAME" VARCHAR(82) NOT NULL, \n\t"NATURE_OF_BUSINESS" VARCHAR(25), \n\t"STREET_NUMBER" VARCHAR(7), \n\t"STREET_NAME" VARCHAR(24), \n\t"STREET_TYPE" VARCHAR(4), \n\t"POST_DIRECTIONAL" VARCHAR(4), \n\t"PHONE" BIGINT, \n\t"PROGRAM_CATEGORY" VARCHAR(4) NOT NULL, \n\t"PROGRAM_CATEGORY_DESCRIPTION" VARCHAR(40), \n\t"INSPECTION_DATE" DATE NOT NULL, \n\t"INSPECTION_TYPE" VARCHAR(32), \n\t"INSPECTION_DESC" VARCHAR(32) NOT NULL, \n\t"SERIAL_NUM" VARCHAR(9) NOT NULL, \n\t"ACTION_CODE" VARCHAR(2) NOT NULL, \n\t"ACTION_DESC" VARCHAR(34) NOT NULL, \n\t"RESULT_CODE" VARCHAR(2) NOT NULL, \n\t"RESULT_DESC" VARCHAR(37) NOT NULL, \n\t"VIOLATION_CODE" VARCHAR(10), \n\t"VIOLATIO

After I create a database, I connect to it.

In [9]:
%sql sqlite:///food.db

'Connected: None@food.db'

I run a COUNT(*) query to verify that all the data has loaded correctly.

In [10]:
%%sql
SELECT COUNT(*)
FROM "food";

Done.


[(17556,)]

## Problem 3 - explore your data (30 points)

Now that the data is loaded, ask some questions of it!  Identify key variables of interest and note their ranges along with other useful descriptive statistics.  Choose and define a few lines of inquiry, execute queries for each, and describe what you find in your result sets.  Use any query techniques we've seen in class, including aggregate functions, transformations, subqueries, or others as appropriate. 

If you find interesting patterns, adding plots to your exploration would be useful.

Insert a combination of code and text/markdown cells here to explore your data.

I first want to take a look at the first 10 rows of the table.

In [11]:
%%sql
SELECT * 
FROM food LIMIT 10;

Done.


[('00BSAY IT WITH CAKE', 50534, '10200 GOLF COURSE RD NW STE B', 'ALBUQUERQUE', 'NM', '87114', 47341, 'OOBSAY IT WITH CAKE', 'CAKE DECORATING', '10200', 'GOLF COURSE', 'RD', 'NW', 5058339563, '0405', 'Food Processor', '2009-10-21', None, 'EMERGENCY RESPONSE', 'DA8009125', '00', 'NON-GRADED', '00', 'NOT APPLICABLE', '68', 'Delivered Informational Brochure', None),
 ('00BSAY IT WITH CAKE', 50534, '10200 GOLF COURSE RD NW STE B', 'ALBUQUERQUE', 'NM', '87114', 47341, 'OOBSAY IT WITH CAKE', 'CAKE DECORATING', '10200', 'GOLF COURSE', 'RD', 'NW', 5058339563, '0405', 'Food Processor', '2009-10-21', None, 'ROUTINE FOOD INSPECTION', 'DA8009123', '11', 'APPROVED', '03', 'IN COMPLIANCE', '00', 'No Violations Found', None),
 ('00BSAY IT WITH CAKE', 50534, '10200 GOLF COURSE RD NW STE B', 'ALBUQUERQUE', 'NM', '87114', 47341, 'OOBSAY IT WITH CAKE', 'CAKE DECORATING', '10200', 'GOLF COURSE', 'RD', 'NW', 5058339563, '0405', 'Food Processor', '2009-10-21', None, 'ROUTINE FOOD INSPECTION', 'DA8009123', '11', 'APPROVED', '03', 'IN COMPLIANCE', '55', 'Additional Comments', None),
 ('00BSAY IT WITH CAKE', 50534, '10200 GOLF COURSE RD NW STE B', 'ALBUQUERQUE', 'NM', '87114', 47341, 'OOBSAY IT WITH CAKE', 'CAKE DECORATING', '10200', 'GOLF COURSE', 'RD', 'NW', 5058339563, '0405', 'Food Processor', '2010-04-14', None, 'ROUTINE FOOD INSPECTION', 'DA8010205', '11', 'APPROVED', '03', 'IN COMPLIANCE', '00', 'No Violations Found', None),
 ('00BSAY IT WITH CAKE', 50534, '10200 GOLF COURSE RD NW STE B', 'ALBUQUERQUE', 'NM', '87114', 47341, 'OOBSAY IT WITH CAKE', 'CAKE DECORATING', '10200', 'GOLF COURSE', 'RD', 'NW', 5058339563, '0405', 'Food Processor', '2010-04-14', None, 'ROUTINE FOOD INSPECTION', 'DA8010205', '11', 'APPROVED', '03', 'IN COMPLIANCE', '55', 'Additional Comments', None),
 ('2000 VIETNAM RESTAURANT', 31468, '601 SAN MATEO BLVD SE', 'ALBUQUERQUE', 'NM', '87108', 29294, '2000 VIETMAN RESTAURANT', 'RESTAURANT', '601', 'SAN MATEO', 'BLVD', 'SE', 5052320900, '0406', 'Food Service Establishment', '2010-03-23', None, 'ROUTINE FOOD INSPECTION', 'DA8509472', '11', 'APPROVED', '03', 'IN COMPLIANCE', '34', 'Food contact surfaces', None),
 ('2000 VIETNAM RESTAURANT', 31468, '601 SAN MATEO BLVD SE', 'ALBUQUERQUE', 'NM', '87108', 29294, '2000 VIETMAN RESTAURANT', 'RESTAURANT', '601', 'SAN MATEO', 'BLVD', 'SE', 5052320900, '0406', 'Food Service Establishment', '2010-03-23', None, 'ROUTINE FOOD INSPECTION', 'DA8509472', '11', 'APPROVED', '03', 'IN COMPLIANCE', '35', 'Non-Food contact surfaces', None),
 ('2000 VIETNAM RESTAURANT', 31468, '601 SAN MATEO BLVD SE', 'ALBUQUERQUE', 'NM', '87108', 29294, '2000 VIETMAN RESTAURANT', 'RESTAURANT', '601', 'SAN MATEO', 'BLVD', 'SE', 5052320900, '0406', 'Food Service Establishment', '2010-03-23', None, 'ROUTINE FOOD INSPECTION', 'DA8509472', '11', 'APPROVED', '03', 'IN COMPLIANCE', '55', 'Additional Comments', None),
 ('328 CHINESE CUISINE', 72789, '5617 MENAUL BLVD NE STE A', 'ALBUQUERQUE', 'NM', '87110', 66644, '328 CHINESE CUISINE', None, '5617', 'MENAUL', 'BLVD', 'NE', 5058814468, '0406', 'Food Service Establishment', '2009-12-15', None, 'EMERGENCY RESPONSE', 'DA7757250', '00', 'NON-GRADED', '00', 'NOT APPLICABLE', '68', 'Delivered Informational Brochure', None),
 ('328 CHINESE CUISINE', 72789, '5617 MENAUL BLVD NE STE A', 'ALBUQUERQUE', 'NM', '87110', 66644, '328 CHINESE CUISINE', None, '5617', 'MENAUL', 'BLVD', 'NE', 5058814468, '0406', 'Food Service Establishment', '2009-12-15', None, 'ROUTINE FOOD INSPECTION', 'DA7757248', '11', 'APPROVED', '03', 'IN COMPLIANCE', '00', 'No Violations Found', None)]

Do the INSPECTION_TYPE column and the INSPECTION_MEMO column have different values other than null values?

In [23]:
%%sql
SELECT DISTINCT INSPECTION_TYPE
FROM food;

Done.


[(None,)]

In [13]:
%%sql
SELECT DISTINCT INSPECTION_MEMO
FROM food;

Done.


[(None,)]

I find the both columns only have null values.

What are different types of actions taken to the inspections?

In [14]:
%%sql
SELECT DISTINCT ACTION_DESC
FROM food;

Done.


[('NON-GRADED',),
 ('APPROVED',),
 ('UNSATISFACTORY',),
 ('ANNUAL STICKER ISSUED',),
 ('NO PERMIT APPLICATION ISSUED',),
 ('ANNUAL STICKER NOT ISSUED',),
 ('FOOD PERMIT SUSPENSION',),
 ('FACILITY CLOSED AT TIME OF SERVICE',)]

How many actions are approved in all these cases?

In [15]:
%%sql
SELECT COUNT (*)
FROM food WHERE ACTION_DESC = "APPROVED";

Done.


[(14080,)]

How many insepections are given to each owner?

In [16]:
%%sql
SELECT OWNER_NAME, COUNT(OWNER_NAME) AS number_of_inspections 
FROM food
GROUP BY OWNER_NAME
ORDER BY COUNT(OWNER_NAME) DESC;

Done.


[('APS', 792),
 ('SMITHS FOOD AND DRUG CENTERS INC', 321),
 ('CIRCLE K STORES INC', 277),
 ('ALBERTSONS', 228),
 ('WAL MART STORES EAST  LP', 156),
 ('BLAKES LOTABURGER LLC', 150),
 ('PALO ALTO INC', 146),
 ('SCS BEVERAGE INC', 138),
 ('WALGREENS CO', 137),
 ('STARBUCKS CORP', 123),
 ('OOBPCF SALESCO LLC', 115),
 ('COA - FAMILY & COMMUNITY SVCS', 114),
 ('PETER DEFRIES CORP', 114),
 ('WESTERN REFINING SOUTHWEST INC', 105),
 ('FAMILY DOLLAR INC 2828', 99),
 ('OVATIONS FOOD SERVICES', 94),
 ('LUIS LUERANO', 85),
 ('OOB GARDUNOS OF MEXICO', 84),
 ('CST SERVICES LLC', 83),
 ('YDI HEAD START', 75),
 ('DOUBLE CHEESE CORP', 69),
 ('ARAMARK ENTERTAINMENT  INC', 68),
 ('ECHO INC', 68),
 ('EVER READY OIL CO INC', 68),
 ('CANTEEN SNACK BAR', 65),
 ('DOLLAR TREE STORES INC', 65),
 ('VERONICA FLORES', 65),
 ('SAMS EAST INC', 64),
 ('TALIN ENTERPRISES INC', 64),
 ('LA PETITE ACADEMY INC', 63),
 ('COSTCO WHOLESALE CORPORATION', 62),
 ('PITAK SIAMSAKOON', 61),
 ('GOLDEN CITIES INC', 57),
 ('ROBERTS OIL', 57),
 ('TWISTERS INC', 57),
 ('SUNFLOWER FARMERS MARKET INC', 54),
 ('BRINKER RESTAURANT CORPORATION', 53),
 ('KEVA JUICE SW LLC', 52),
 ('PROS ABQ RANCH MARKET LLC', 52),
 ('ERIC KILMER', 51),
 ('PAND ENTERPRISES INC', 51),
 ('G & J BROOKS ENTERPRISES  INC', 48),
 ('ABE GARCIA/DENISE', 46),
 ('GMRI INC', 46),
 ('TARGET CORPORATION', 46),
 ('LA HACIENDA OF OLD TOWN INC', 45),
 ('WHATABURGER RESTAURANTS LP', 45),
 ('ADOLFO AQUILAR RAMIREZ', 43),
 ('SERVICE SYSTEMS ASSOC.INC.', 42),
 ('FRESQUEZ CONCESSIONS INC', 41),
 ('STAR OF THE HIGH DESERT INC', 41),
 ('COA - DEPT OF SENIOR AFFAIRS', 40),
 ('DCT ENTERPRISES OF NM LLC', 39),
 ('FLYING STAR CAFE INC', 39),
 ('GASS INC', 39),
 ('EINSTEIN AND NOAH CORP', 36),
 ('EL MEZQUITE MARKET LLC', 36),
 ('WALKING BY FAITH INC', 34),
 ('LPG ENTERPRISES INC', 33),
 ('SANTA FE PICACHO MGMT', 33),
 ('CAJUN OPERATING COMPANY', 32),
 ('FRANS ENTERPRISES', 32),
 ('HASTINGS ENTERTAINMENT INC', 32),
 ('PANDA EXPRESS INC', 32),
 ('GARCIAS KITCHENS', 31),
 ('RUTH WENDELL', 31),
 ('SOUPER SALAD  INC. #38', 31),
 ('BURGER KING', 30),
 ('GENERAL NUTRITION CORPORATION', 30),
 ('AMERICAN BLUE RIBBON HOLDINGS', 29),
 ('BENCO FOODS INC', 29),
 ('CLIFFS', 29),
 ('IVAN VERTIZ', 29),
 ('LITTLE ANITAS MEXICAN FOOD INC', 29),
 ('SMOKIN DJS', 29),
 ('SUSHI KING LLC', 29),
 ('KINDERCARE LEARING CENTER', 28),
 ('ALBUQ ENTERTAINMENT COMPLEX', 27),
 ('BOBA TEA COMPANY INC', 27),
 ('MJG CORPORATION', 27),
 ('MMRV LLC', 27),
 ('NELLOS BROTHERS INC', 27),
 ('WILLIS INVESTMENT CORPORATION', 27),
 ('DUKE CITY PIZZA INC', 26),
 ('HI LO MARKET', 26),
 ('LITTLE CAESARS ASF CORP', 26),
 ('PAY-N-SAVE  INC.', 26),
 ('WHOLE FOODS MARKET', 26),
 ('AMERICAN GOLF CORP.', 25),
 ('OOB 10.30.13 QK NEW MEX INC', 25),
 ('OOB J & K LLC', 25),
 ('OOB SHARMIN KHANDAKER', 25),
 ('BERMUDEZ BROS LLC', 24),
 ('BIG 5 CORP', 24),
 ('CONQUISTADORES INC #4', 24),
 ('DOLGEN CORP INC', 24),
 ('KENRIC INC', 24),
 ('OOB PER WRITTEN REQUEST 12/23/13 ANK CORP', 24),
 ('RESTAURANTS OF AMERICA INC', 24),
 ('SONIC DRIVE IN', 24),
 ('LA MONTANITA FOOD CO OP', 23),
 ('NM CONGRESS OF PTA', 23),
 ('NORTHERN NEW MEXICO FOODS INC', 23),
 ('R M C/ A S P', 23),
 ('ARDENT HEALTH SERVICES', 22),
 ('G & P INVESTMENTS INC', 22),
 ('HYATT REGENCY ALBUQUERQUE', 22),
 ('LEOPOLDO NUNEZ', 22),
 ('PETE RALLIS', 22),
 ('REGAL CINEMAS INC', 22),
 ('SATELLITE COFFEE  INC', 22),
 ('TIFFANY TRUONG', 22),
 ('ADOLFO AQUILAR', 21),
 ('ADVANCED FRESH CONCEPTS FRANCHISE CORP', 21),
 ('FILO LLC', 21),
 ('LQ MANAGEMENT LLC', 21),
 ('OOBTESSTAB ENTERPRISES INC', 21),
 ('RIO GRANDE PIZZA INC', 21),
 ('WESTGATE LITTLE LEAGUE', 21),
 ('6001 INC', 20),
 ('ALLSUPS CONVENIENCE STORES', 20),
 ('DEL NORTE HIGH SCHOOL', 20),
 ('MANZANO HIGH SCHOOL', 20),
 ('THE HOME DEPOT - TAX DEPT/L&P', 20),
 ('ARTURO VEGA', 19),
 ('CHICADEES', 19),
 ('KELLERS FARM STORE', 19),
 ('OOB CLUB CHIENNE', 19),
 ('TWIN DRAGONS INC', 19),
 ('ALBUQ COUNTRY CLUB', 18),
 ('BIRD OF PARADISE  INC', 18),
 ('FRONTIER RESTAURANT', 18),
 ('JE

I find APS gets most inspections and let's take a close look.

In [17]:
%%sql
SELECT INSPECTION_DESC,
COUNT(INSPECTION_DESC) FROM food AS food
WHERE OWNER_NAME = 'APS'
GROUP BY "INSPECTION_DESC"
ORDER BY COUNT(INSPECTION_DESC) DESC;

Done.


[('ROUTINE FOOD INSPECTION', 589),
 ('EMERGENCY RESPONSE', 118),
 ('MOBILE UNIT FOOD INSPECTION', 32),
 ('New Business Inspection', 24),
 ('PRE-OPENING FOOD INSPECTION', 14),
 ('ADDITIONAL SERVICE PROVIDED', 6),
 ('FOLLOW-UP DOCUMENTATION RECEIVED', 5),
 ('FOOD FOLLOW-UP INSPECTION', 2),
 ('MANAGERIAL ASSISTANCE', 1),
 ('PHONE CALL', 1)]

From above, it can be found that APS gets a large number of routine food inspections and it also gets many new business inspection and pre-opening food inspections. It appears that APS is newly opened and that may be the reason why it gets most inspections.

What are different types of violation and which one is the most popular one?

In [18]:
%%sql
SELECT DISTINCT VIOLATION_CODE, COUNT(*)
FROM food
GROUP BY VIOLATION_CODE
ORDER BY COUNT(*) DESC;

Done.


[('55', 6348),
 ('00', 3429),
 ('68', 1658),
 (None, 1009),
 ('42', 487),
 ('35', 441),
 ('21', 429),
 ('38', 420),
 ('45', 401),
 ('34', 246),
 ('01', 196),
 ('36', 178),
 ('02', 170),
 ('46', 137),
 ('40', 129),
 ('22', 119),
 ('09', 109),
 ('31', 108),
 ('39', 92),
 ('51', 91),
 ('32', 88),
 ('19', 84),
 ('33', 75),
 ('37', 58),
 ('58', 56),
 ('05', 52),
 ('43', 44),
 ('48', 43),
 ('03', 40),
 ('24', 39),
 ('16', 37),
 ('26', 37),
 ('08', 36),
 ('07', 35),
 ('25', 34),
 ('10', 28),
 ('13', 23),
 ('47', 20),
 ('44', 19),
 ('27', 18),
 ('41', 17),
 ('15', 16),
 ('11', 15),
 ('23', 15),
 ('57', 15),
 ('17', 13),
 ('20', 13),
 ('12', 11),
 ('30', 11),
 ('14', 9),
 ('52', 9),
 ('49', 7),
 ('61', 7),
 ('59', 6),
 ('04  00', 5),
 ('04  01', 5),
 ('04  02', 5),
 ('04  03', 5),
 ('04  04', 5),
 ('04  05', 5),
 ('04  06', 5),
 ('04  07', 5),
 ('04  08', 5),
 ('04  09', 5),
 ('04  10', 5),
 ('04  11', 5),
 ('04  12', 5),
 ('04  13', 5),
 ('04  14', 5),
 ('04  15', 5),
 ('04  16', 5),
 ('04  17', 5),
 ('04  18', 5),
 ('04  19', 5),
 ('04  20', 5),
 ('04  21', 5),
 ('04  22', 5),
 ('04  23', 5),
 ('04  24', 5),
 ('04  25', 5),
 ('04  26', 5),
 ('04  27', 5),
 ('04  28', 5),
 ('04  29', 5),
 ('04  30', 5),
 ('04  31', 5),
 ('04  32', 5),
 ('04  33', 5),
 ('04  34', 5),
 ('04  35', 5),
 ('04  36', 5),
 ('04  37', 5),
 ('04  38', 5),
 ('04  39', 5),
 ('04  40', 5),
 ('04  41', 5),
 ('04  42', 5),
 ('04  43', 5),
 ('04  44', 5),
 ('04  45', 5),
 ('04  46', 5),
 ('04  47', 5),
 ('04  48', 5),
 ('04  49', 5),
 ('04  50', 5),
 ('04  51', 5),
 ('04  58', 5),
 ('04  59', 5),
 ('04  60', 5),
 ('04  61', 5),
 ('04  69', 5),
 ('04  70', 5),
 ('04  71', 5),
 ('04  72', 5),
 ('04  73', 5),
 ('50', 5),
 ('28', 4),
 ('60', 3),
 ('29', 2),
 ('02 15A0204', 1),
 ('04', 1),
 ('06', 1),
 ('07  D1', 1),
 ('07  H1', 1),
 ('07  H4', 1),
 ('56', 1),
 ('66', 1),
 ('\\', 1),
 (']', 1)]

It can be found that 55 is the most frequent violation type.

How many different types of inspections are taken each day per type?

In [19]:
%%sql
SELECT INSPECTION_DATE, INSPECTION_DESC, COUNT(INSPECTION_DESC)
FROM food
GROUP BY INSPECTION_DATE;

Done.


[('2009-07-01', 'PRE-OPENING FOOD INSPECTION', 34),
 ('2009-07-02', 'ROUTINE FOOD INSPECTION', 75),
 ('2009-07-06', 'ROUTINE FOOD INSPECTION', 89),
 ('2009-07-07', 'ROUTINE FOOD INSPECTION', 60),
 ('2009-07-08', 'ROUTINE FOOD INSPECTION', 58),
 ('2009-07-09', 'ROUTINE FOOD INSPECTION', 94),
 ('2009-07-10', 'ROUTINE FOOD INSPECTION', 83),
 ('2009-07-11', 'FOLLOW-UP DOCUMENTATION RECEIVED', 1),
 ('2009-07-13', 'ROUTINE FOOD INSPECTION', 77),
 ('2009-07-14', 'ROUTINE FOOD INSPECTION', 100),
 ('2009-07-15', 'ROUTINE FOOD INSPECTION', 65),
 ('2009-07-16', 'ROUTINE FOOD INSPECTION', 126),
 ('2009-07-17', 'ROUTINE FOOD INSPECTION', 57),
 ('2009-07-18', 'ROUTINE FOOD INSPECTION', 1),
 ('2009-07-20', 'ROUTINE FOOD INSPECTION', 82),
 ('2009-07-21', 'ROUTINE FOOD INSPECTION', 115),
 ('2009-07-22', 'ROUTINE FOOD INSPECTION', 76),
 ('2009-07-23', 'ROUTINE FOOD INSPECTION', 79),
 ('2009-07-24', 'ROUTINE FOOD INSPECTION', 69),
 ('2009-07-27', 'ROUTINE FOOD INSPECTION', 86),
 ('2009-07-28', 'New Business Inspection', 105),
 ('2009-07-29', 'ROUTINE FOOD INSPECTION', 54),
 ('2009-07-30', 'ROUTINE FOOD INSPECTION', 65),
 ('2009-07-31', 'ROUTINE FOOD INSPECTION', 42),
 ('2009-08-01', 'ROUTINE FOOD INSPECTION', 1),
 ('2009-08-03', 'ROUTINE FOOD INSPECTION', 51),
 ('2009-08-04', 'ROUTINE FOOD INSPECTION', 46),
 ('2009-08-05', 'ROUTINE FOOD INSPECTION', 58),
 ('2009-08-06', 'ROUTINE FOOD INSPECTION', 67),
 ('2009-08-07', 'ROUTINE FOOD INSPECTION', 92),
 ('2009-08-10', 'ROUTINE FOOD INSPECTION', 91),
 ('2009-08-11', 'PRE-OPENING FOOD INSPECTION', 78),
 ('2009-08-12', 'ROUTINE FOOD INSPECTION', 64),
 ('2009-08-13', 'ROUTINE FOOD INSPECTION', 24),
 ('2009-08-14', 'ROUTINE FOOD INSPECTION', 44),
 ('2009-08-17', 'ROUTINE FOOD INSPECTION', 75),
 ('2009-08-18', 'ROUTINE FOOD INSPECTION', 75),
 ('2009-08-19', 'ROUTINE FOOD INSPECTION', 53),
 ('2009-08-20', 'ROUTINE FOOD INSPECTION', 65),
 ('2009-08-21', 'ROUTINE FOOD INSPECTION', 67),
 ('2009-08-24', 'ROUTINE FOOD INSPECTION', 82),
 ('2009-08-25', 'ROUTINE FOOD INSPECTION', 84),
 ('2009-08-26', 'ADDITIONAL SERVICE PROVIDED', 65),
 ('2009-08-27', 'ROUTINE FOOD INSPECTION', 100),
 ('2009-08-28', 'ROUTINE FOOD INSPECTION', 78),
 ('2009-08-31', 'ROUTINE FOOD INSPECTION', 80),
 ('2009-09-01', 'ROUTINE FOOD INSPECTION', 124),
 ('2009-09-02', 'EMERGENCY RESPONSE', 118),
 ('2009-09-03', 'EMERGENCY RESPONSE', 115),
 ('2009-09-04', 'ROUTINE FOOD INSPECTION', 72),
 ('2009-09-08', 'EMERGENCY RESPONSE', 80),
 ('2009-09-09', 'ROUTINE FOOD INSPECTION', 52),
 ('2009-09-10', 'ROUTINE FOOD INSPECTION', 159),
 ('2009-09-11', 'EMERGENCY RESPONSE', 101),
 ('2009-09-14', 'ROUTINE FOOD INSPECTION', 67),
 ('2009-09-15', 'ROUTINE FOOD INSPECTION', 129),
 ('2009-09-16', 'PRE-OPENING FOOD INSPECTION', 75),
 ('2009-09-17', 'ROUTINE FOOD INSPECTION', 106),
 ('2009-09-18', 'ROUTINE FOOD INSPECTION', 71),
 ('2009-09-21', 'New Business Inspection', 79),
 ('2009-09-22', 'PRE-OPENING FOOD INSPECTION', 106),
 ('2009-09-23', 'ROUTINE FOOD INSPECTION', 79),
 ('2009-09-24', 'ROUTINE FOOD INSPECTION', 92),
 ('2009-09-25', 'MANAGERIAL ASSISTANCE', 148),
 ('2009-09-28', 'ROUTINE FOOD INSPECTION', 81),
 ('2009-09-29', 'ROUTINE FOOD INSPECTION', 93),
 ('2009-09-30', 'ROUTINE FOOD INSPECTION', 83),
 ('2009-10-01', 'ROUTINE FOOD INSPECTION', 131),
 ('2009-10-02', 'ROUTINE FOOD INSPECTION', 49),
 ('2009-10-05', 'ROUTINE FOOD INSPECTION', 60),
 ('2009-10-06', 'New Business Inspection', 34),
 ('2009-10-07', 'ROUTINE FOOD INSPECTION', 63),
 ('2009-10-08', 'ROUTINE FOOD INSPECTION', 58),
 ('2009-10-09', 'EMERGENCY RESPONSE', 71),
 ('2009-10-12', 'ROUTINE FOOD INSPECTION', 42),
 ('2009-10-13', 'ROUTINE FOOD INSPECTION', 79),
 ('2009-10-14', 'ROUTINE FOOD INSPECTION', 90),
 ('2009-10-15', 'ROUTINE FOOD INSPECTION', 121),
 ('2009-10-16', 'ROUTINE FOOD INSPECTION', 52),
 ('2009-10-19', 'ROUTINE FOOD INSPECTION', 101),
 ('2009-10-20', 'ROUTINE FOOD INSPECTION', 107),
 ('2009-10-21', 'ROUTINE FOOD INSPECTION', 100),
 ('2009-10-22', 'PRE-OPENING FOOD INSPECTION',

## Bonus (10 points)

We have seen a few simple techniques for normalizing data in class.  Choose a column or set of columns that might be redundant and normalize them out to another table:

* Identify the columns to normalize in the source data table
* Create one or more new tables to hold the normalized records
* Extract data into the new tables
* Re-insert the new foreign key references from the new table into the source table

I would like to normalize violation code and violation description.

First, I create a table named violations.

In [20]:
%%sql
DROP TABLE IF EXISTS violations;
CREATE TABLE violations
(code int primary key,
description varchar(64)
);

Done.
Done.


[]

In [21]:
%%sql
INSERT INTO violations (code)
SELECT DISTINCT CAST("VIOLATION_CODE" as int) FROM food;

62 rows affected.


[]

In [22]:
%%sql
UPDATE violations as v
SET description = f."VIOLATION_DESC"
FROM food as f
WHERE code = CAST(f.VIOLATION_CODE AS int);

(sqlite3.OperationalError) near "as": syntax error [SQL: 'UPDATE violations as v\nSET description = f."VIOLATION_DESC"\nFROM food as f\nWHERE code = CAST(f.VIOLATION_CODE AS int);']
